File: Applied Course/4.Problem Solving/6.Problems on Heap/5.K Closest points to origin.cpp
#include<bits/stdc++.h>
using namespace std;
// Logic-1
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        vector<vector<int>> v;
        sort(points.begin(),points.end(),[](vector<int> v1,vector<int> v2){return (v1[0]*v1[0]+v1[1]*v1[1])<(v2[0]*v2[0]+v2[1]*v2[1]);});
        for(int i=0;i<k;i++)
            v.push_back(points[i]);
        return v;
    }
};
// Logic-2
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        vector<vector<int>> v;
        priority_queue<pair<int,vector<int>>> max_heap;
        for(int i=0;i<points.size();i++)
        {
            max_heap.push({points[i][0]*points[i][0]+points[i][1]*points[i][1],points[i]});
            if(max_heap.size()>k)
                max_heap.pop();
        }
        while(max_heap.size())
        {
            v.push_back(max_heap.top().second);
            max_heap.pop();
        }
        return v;
    }
};
int main()
{
    // https://leetcode.com/problems/k-closest-points-to-origin
    return 0;
}

File: Applied Course/4.Problem Solving/6.Problems on Heap/5.K Closest points to origin.py
# https://leetcode.com/problems/k-closest-points-to-origin
# Logic-1
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        points.sort(key=lambda P:P[0]**2+P[1]**2)
        return points[:k]
# Logic-2
import heapq
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        getDist=lambda p:-(p[0]**2+p[1]**2)
        heap=[]
        for point in points:
            if len(heap)<k:
                heapq.heappush(heap,(getDist(point),point))
            else:
                heapq.heappushpop(heap,(getDist(point),point))
        return [i[1] for i in heap]